coresimd/wasm32/simd128.rs

```diff
@@ -0,0 +1,805 @@
+//! This module implements the [WebAssembly `SIMD128` ISA].
+//!
+//! [WebAssembly `SIMD128` ISA]:
+//! https://github.com/WebAssembly/simd/blob/master/proposals/simd/SIMD.md
+
+#![allow(non_camel_case_types)]
+
+/// A single unconstrained byte (0-255).
+#[derive(Copy, Clone, Debug)]
+pub struct ImmByte(u8);
+impl ImmByte {
+    /// Constructor
+    #[inline]
+    #[rustc_args_required_const(0)]
+    pub const fn new(value: u8) -> Self {
+        ImmByte(value)
+    }
+}
+
+macro_rules! impl_laneidx {
+    ($id:ident($ty:ty): [$_from:expr, $_to:expr] | $(#[$doc:meta])*) => {
+        #[derive(Copy, Clone, Debug)]
+        pub struct $id($ty);
+        impl $id {
+            #[inline]
+            #[rustc_args_required_const(0)]
+            pub const fn new(x: $ty) -> Self {
+                // FIXME: not allowed in const fn:
+                // * if statements
+                // * unreachable_unchecked / panic / abort
+                //
+                // if x < $from || x > $to {
+                //     unsafe { ::_core::hint::unreachable_unchecked() };
+                //     debug_assert!(...)
+                // }
+                $id(x)
+            }
+        }
+    };
+}
+impl_laneidx!(LaneIdx2(u8): [0, 1] | /// A byte with values in the range 0–1 identifying a lane.
```

Comment:
> I think it'd be really hard for us to stabilize these LandIdx* types unfortunately,

Why do you say so? 

> would it be possible though to say it's UB to specify an out of bounds index and just take usize for indices?

Duh, the docs don't say it but stating that if the value is out-of-bounds the behavior is undefined was the intent.

I went with the new types because it felt more conservative, and it should allow us to insert debug asserts in the future (undefined behavior here means we can do anything, including properly diagnosing the error at run-time). 

Having been using these a bit while running the tests, I always just ended up transmuting `u8`s anyways, so I am myself unconvinced that this is worth it. (the spec says these are byte sized, so I think it makes sense to just make these `u8` instead of `usize` here)

If we make them newtypes these should all be `#[repr(transparent)]` though as @cramertj pointed out.